### PR TITLE
feat: let agent retry to connecting to server

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/woodpecker-ci/woodpecker/issues/1721

PR introduces connection retries during agent startup.

### CODE REVIEW

File: cmd/agent/main.go [hunk 1]
```diff
@@ -17,19 +17,39 @@ package main
...
+	retryDelay = 2 * time.Second
+)
+
 func main() {
 	app := cli.NewApp()
 	app.Name = "woodpecker-agent"
 	app.Version = version.String()
 	app.Usage = "woodpecker agent"
-	app.Action = loop
+	app.Action = func(context *cli.Context) error {
```
Comment (6543) [inline cmd/agent/main.go:41]: can we move that int it's own func and rename loop to something more descriptive ?

File: cmd/agent/main.go [hunk 2]
```diff
@@ -17,19 +17,39 @@ package main
...
 	"github.com/urfave/cli/v2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 
 	"github.com/woodpecker-ci/woodpecker/version"
 )
 
+const (
+	retryCount = 5
+	retryDelay = 2 * time.Second
```
[the remark this answers is not in the document]
Comment (anbraten) [reply cmd/agent/main.go:33]: I would prefer good common values and not make it configurable to keep config options and therefore miss-configurations minimal.

[the remark this answers is not in the document]
Comment (6543) [reply cmd/agent/main.go:33]: I'm not against make things configurable ... just have good defaults so you dont have to have the "configuration fear" ... and we can put souch config options into an "advanced" chapter

[the remark this answers is not in the document]
Comment (anbraten) [reply cmd/agent/main.go:33]: Anyways why shouldn't the agent try to re-connect infinitely?

File: cmd/agent/agent.go [hunk 3]
```diff
@@ -246,6 +248,19 @@ func loop(c *cli.Context) error {
 	return nil
 }
 
+func rinWithRetry(context *cli.Context) error {
```
Comment (anbraten) [inline cmd/agent/agent.go:251]:
```suggestion
func runWithRetry(context *cli.Context) error {
```

Comment (alexmt) [reply cmd/agent/agent.go]: oh, sorry - fixed

File: cmd/agent/agent.go [hunk 4]
```diff
@@ -246,6 +248,19 @@ func loop(c *cli.Context) error {
...
 
+func runWithRetry(context *cli.Context) error {
+	var err error
+	for i := 0; i < retryCount; i++ {
+		if err = run(context); err == nil {
+			break
+		} else if status.Code(err) == codes.Unavailable {
+			log.Warn().Err(err).Msg(fmt.Sprintf("cannot connect to server, retrying in %v", retryDelay))
+			time.Sleep(retryDelay)
+		}
```
Comment (6543) [inline cmd/agent/agent.go:259]: if err != nil and no rpc error we should break and return err too

Comment (alexmt) [reply cmd/agent/agent.go:261]: Sorry, looks like my brain was not working well. Fixed.

File: cmd/agent/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexmt) [reply cmd/agent/main.go]: Implemented config changes. 5 retries with 2 seconds delay looks reasonable to me so I've used these values.

I spent some time exploring the structure of the docs, trying to find where to fit advanced. The [15-agent-config.md](https://github.com/woodpecker-ci/woodpecker/blob/master/docs/docs/30-administration/15-agent-config.md) already has the most important flags at the top, and the rest of the settings are in `All agent configuration options`. 

The `All agent configuration options` section look `advanced` to me and is a good place to document new retry options. For the sake of time, put new options where but happy to move it if you think differently.

[the remark this answers is not in the document]
Comment (anbraten) [reply cmd/agent/main.go]: Just as I am still not convinced. Why would a user need to change those values? I would still prefer to keep things simple and only add this config option if we get actual feedback / issues where this is really necessary.

[the remark this answers is not in the document]
Comment (6543) [reply cmd/agent/main.go]: K8s is a valid reason i would say :)

**Review by 6543 — APPROVED** [review]

**Pull request merged**